### PR TITLE
Add terminate protection to master instance

### PR DESCRIPTION
**Pull request by pipo02mix**
Towards https://github.com/giantswarm/giantswarm/issues/5453

### CODE REVIEW

**Comment by pipo02mix** [comment]
Tested on ginger and works fine (upgrade too)

File: service/controller/v24/resource/cloudformation/delete.go [hunk 1]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+				},
+			},
+		},
+	}
+	result, err := cc.AWSClient.EC2.DescribeInstances(input)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, reservation := range result.Reservations {
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]: For now we only have one master and we should verify we only modify one. When there are more masters in the result, or fewer, we should return an error.

File: service/controller/v24/resource/cloudformation/delete.go [hunk 2]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
 	return nil
 }
 
+func disableMasterTerminationProtection(ctx context.Context, masterInstanceName string) error {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	input := &ec2.DescribeInstancesInput{
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]:
```suggestion
	i := &ec2.DescribeInstancesInput{
```

File: service/controller/v24/resource/cloudformation/delete.go [hunk 3]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("tag:Name"),
+				Values: []*string{
+					aws.String(masterInstanceName),
+				},
+			},
+		},
+	}
+	result, err := cc.AWSClient.EC2.DescribeInstances(input)
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]:
```suggestion
	o, err := cc.AWSClient.EC2.DescribeInstances(i)
```

File: service/controller/v24/resource/cloudformation/delete.go (hunk 1 above)
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]:
```suggestion
	for _, reservation := range o.Reservations {
```

File: service/controller/v24/resource/cloudformation/delete.go [hunk 4]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, reservation := range result.Reservations {
+		for _, instance := range reservation.Instances {
+
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]:
```suggestion
```

File: service/controller/v24/resource/cloudformation/delete.go [hunk 5]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+			input := &ec2.ModifyInstanceAttributeInput{
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]:
```suggestion
			i := &ec2.ModifyInstanceAttributeInput{
```

File: service/controller/v24/resource/cloudformation/delete.go [hunk 6]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+				DisableApiTermination: &ec2.AttributeBooleanValue{
+					Value: aws.Bool(false),
+				},
+				InstanceId: aws.String(*instance.InstanceId),
+			}
+
+			_, err = cc.AWSClient.EC2.ModifyInstanceAttribute(input)
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]:
```suggestion
			_, err = cc.AWSClient.EC2.ModifyInstanceAttribute(i)
```

File: service/controller/v24/resource/cloudformation/delete.go [hunk 7]
```diff
@@ -100,6 +108,47 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
 	return nil
 }
 
+func disableMasterTerminationProtection(ctx context.Context, masterInstanceName string) error {
+	cc, err := controllercontext.FromContext(ctx)
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]: Can we have start and end logs in here? Basically `disabling master instance termination protection` and `disabled master instance termination protection`.

File: service/controller/v24/resource/cloudformation/delete.go [hunk 8]
```diff
@@ -31,6 +32,13 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+		err = disableMasterTerminationProtection(ctx, key.MasterInstanceName(customObject))
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "disabled master instance termination protection")
+
 		stackName := aws.String(key.MainGuestStackName(customObject))
 
 		updateTerminationProtection := &cloudformation.UpdateTerminationProtectionInput{
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]: What is this doing then?

Comment (pipo02mix) [reply service/controller/v24/resource/cloudformation/delete.go]: One is for the cloud formation stack and the other for the master instance (EC2 instance)

Comment (xh3b4sd) [reply service/controller/v24/resource/cloudformation/delete.go]: 💡

File: service/controller/v24/resource/cloudformation/delete.go [hunk 9]
```diff
@@ -100,6 +106,57 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+		return microerror.Mask(err)
+	}
+
+	for _, reservation := range o.Reservations {
+
+		if len(reservation.Instances) < 1 {
+			return microerror.Mask(notMasterInstanceExistsError)
+		} else if len(reservation.Instances) > 1 {
+			return microerror.Mask(moreThanOneMasterInstanceExistsError)
+		}
```
Comment (pipo02mix) [inline service/controller/v24/resource/cloudformation/delete.go]: I am not sure if there is a nicer solution

Comment (xh3b4sd) [reply service/controller/v24/resource/cloudformation/delete.go]: We can do it just like here: https://github.com/giantswarm/aws-operator/blob/master/service/routetable/route_table.go#L91-L93.

**Comment by pipo02mix** [comment]
done PTAL

File: service/controller/v24/resource/cloudformation/delete.go [hunk 10]
```diff
@@ -69,6 +75,55 @@ func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange inte
...
+	}
+	o, err := cc.AWSClient.EC2.DescribeInstances(i)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, reservation := range o.Reservations {
+
+		if len(reservation.Instances) != 1 {
+			return microerror.Maskf(masterInstanceExistsError, "expected one master instance, got %d", len(reservation.Instances))
```
Comment (xh3b4sd) [inline service/controller/v24/resource/cloudformation/delete.go]: A bit nitty but the error is not really accurate. What about some generic one like `executionFailedError`?

Comment (xh3b4sd) [reply service/controller/v24/resource/cloudformation/delete.go]: Also, how many reservations are there? We should also check that there is only one of them.

Comment (rossf7) [reply service/controller/v24/resource/cloudformation/delete.go]: Good point. We could return a `tooManyError` in that case.

Comment (xh3b4sd) [reply service/controller/v24/resource/cloudformation/delete.go]: When you want to do that we have to check twice. I found that cumbersome in some other places and just checked that there is one item and returned the generic error.

Comment (rossf7) [reply service/controller/v24/resource/cloudformation/delete.go]: Ack that was the previous approach. Agree it is more cumbersome. Lets go with this check and the generic error.

**Comment by pipo02mix** [comment]
> Also, how many reservations are there? We should also check that there is only one of them.

Reading the docs:
```
A reservation corresponds to a command to start instances. You can see what instances are associated with a reservation
```
So it is the action of launch instance. In our case, it should be only one but it is managed by CF AFAIU

**Comment by pipo02mix** [comment]
Tested and works fine PTAL

**Comment by xh3b4sd** [comment]
I think this change broke upgrades from v24 to v25. 

> {"caller":"github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/operatorkit/controller/resource/retryresource/crud_resource_ops_wrapper.go:213","controller":"aws-operator","event":"update","function":"ApplyUpdateChange","level":"warning","loop":"318","message":"retrying due to error","object":"/apis/provider.giantswarm.io/v1alpha1/namespaces/default/awsconfigs/0ytga","resource":"tccpv25","stack":"[{/go/src/github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/operatorkit/controller/resource/retryresource/crud_resource_ops_wrapper.go:206: } {/go/src/github.com/giantswarm/aws-operator/service/controller/v25/resource/tccp/update.go:80: } {/go/src/github.com/giantswarm/aws-operator/service/controller/v25/resource/tccp/update.go:393: } {OperationNotPermitted: The instance 'i-01c4e0c909f45cc1f' may not be terminated. Modify its 'disableApiTermination' instance attribute and try again.\n\tstatus code: 400, request id: 9eedd16d-e3b7-44ed-b069-ab07674d2f9d}]","time":"2019-03-15T10:02:27.467394+00:00","underlyingResource":"tccpv25","version":"32377452"}

**Pull request merged**